src/renderer/main.tsx

```diff
@@ -61,9 +62,37 @@ export const Main = memo(() => {
         )
     );
 
+    const [status, setStatus] = useState(t('splash.loading', 'Loading...'));
+    const [statusProgress, setStatusProgress] = useState<null | number>(null);
+    const [overallProgress, setOverallProgress] = useState(0);
+
+    // Register event listeners
+    useEffect(() => {
+        ipcRenderer.on('setup-progress', (event, progress) => {
+            setOverallProgress(progress.totalProgress);
+            setStatusProgress(progress.statusProgress > 0 ? progress.statusProgress : null);
+
+            if (progress.status) {
+                setStatus(progress.status);
+            }
+        });
+    }, []);
+
+    useEffect(() => {
+        ipcRenderer.send(isBackendReady ? 'enable-menu' : 'disable-menu');
+        if (isBackendReady) {
+            setOverallProgress(1);
+        }
+    }, [isBackendReady]);
+
     if (connectionState === 'failed') return null;
 
-    if (connectionState === 'connecting' || !ready) {
+    if (
+        connectionState === 'connecting' ||
+        !ready ||
+        !isBackendReady ||
+        schemata.schemata.length === 0
```

Comment:
> schemata.schemata.length === 0

This is weird. Is this to catch the case of a fresh install when the worker doesn't have any nodes for a few seconds?

Reply:
There's some kind of race condition where the backend is ready but we haven't fetched the nodes yet. At least, there was. It's possible with my changes that this line is no longer necessary, but previously it was trying to get the first schema item and failing, so I added this

Reply:
So the issue isn't with the frontend, but that the host server passes `/nodes` requests along to the worker before the backend is actually ready.

Let's leave this hack in for now. I already have an idea how to fix this.

Reply:
lol

Reply:
oh whoops, i just removed it. in my latest testing it wasnt necessary anymore

Reply:
It is still necessary, but don't worry about it. I'll fix it after this PR.